libs/pika/executors/include/pika/executors/limiting_scheduler.hpp

```diff
@@ -0,0 +1,190 @@
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <pika/config.hpp>
+#include <pika/assert.hpp>
+#include <pika/coroutines/thread_enums.hpp>
+#include <pika/debugging/print.hpp>
+#include <pika/errors/try_catch_exception_ptr.hpp>
+#include <pika/execution/algorithms/execute.hpp>
+#include <pika/execution/algorithms/schedule_from.hpp>
+#include <pika/execution/executors/execution_parameters.hpp>
+#include <pika/execution_base/receiver.hpp>
+#include <pika/execution_base/sender.hpp>
+#include <pika/synchronization/counting_semaphore.hpp>
+#include <pika/threading_base/annotated_function.hpp>
+#include <pika/threading_base/register_thread.hpp>
+#include <pika/threading_base/thread_description.hpp>
+
+#include <cstddef>
+#include <exception>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+namespace pika::execution::experimental {
+
+    // -----------------------------------------------------------------
+    // by convention the title is 7 chars (for alignment)
+    // a debug level of N shows messages with level 1..N
+    using namespace pika::debug::detail;
+    template <int Level>
+    static print_threshold<Level, 1> lsc_debug("SCLIMIT");
+
+    template <typename WrappedScheduler>
+    struct limiting_scheduler
+    {
+        using semaphore_type = pika::counting_semaphore<>;
+        //
+        constexpr limiting_scheduler() = delete;
+        explicit limiting_scheduler(int max_n, WrappedScheduler on_scheduler)
+          : internal_scheduler_(on_scheduler)
+        {
+            semaphore_ = std::make_shared<semaphore_type>(max_n);
+            PIKA_DETAIL_DP(lsc_debug<5>, debug(str<>("construct"), max_n));
+        }
+
+        /// \cond NOINTERNAL
+        bool operator==(limiting_scheduler const& rhs) const noexcept
+        {
+            return internal_scheduler_ == rhs.internal_scheduler_;
+        }
+
+        bool operator!=(limiting_scheduler const& rhs) const noexcept
+        {
+            return !(*this == rhs);
+        }
+
+        WrappedScheduler& get_internal_scheduler()
+        {
+            return internal_scheduler_;
+        }
+
+        template <typename F>
+        void execute(F&& f) const
+        {
+            using namespace pika::debug::detail;
+            PIKA_DETAIL_DP(lsc_debug<5>, debug(str<>("acquire")));
+            semaphore_->acquire();
+            auto semaphore_lambda = [f = std::forward<F>(f), sem = this->semaphore_]() mutable {
+                PIKA_INVOKE(f, );
+                PIKA_DETAIL_DP(lsc_debug<5>, debug(str<>("release")));
+                sem->release();
+            };
+            pika::execution::experimental::execute(
+                internal_scheduler_, PIKA_MOVE(semaphore_lambda));
+        }
+
+        template <typename F>
+        friend void tag_invoke(execute_t, limiting_scheduler const& sched, F&& f)
+        {
+            execute(sched, PIKA_FORWARD(F, f));
```

Comment:
Oh right, this needs to be
```suggestion
            sched.execute(PIKA_FORWARD(F, f));
```
Sorry for the previous suggestion...